New form home page output

fkang-pu opened the pull request:
**Summaries:**

I have made significant changes to form, views, and templates. 

- Use "bootstrap-select" plugin for selecting a list of options for stim_neu_id and resp_neu_ids.
   It works better with Django forms than "vanillaSelectBox" plugin which was used previously.

-  Created a Python class in utils.py to call methods in wormfunconn package and generate html_div for plot(s). Also capture possible error(s) and write to a dictionary instead of raise error(s). It makes the code in views.py simple and easy to follow, and allows displaying form or application error(s) on web page.

- The neural response simulator code calls different sets of parameters based on stim_type. The design is to list all parameters in a Django form, but only display required parameters on web form based on selected stim_type. The hidden fields are set to initial values by JavaScript.

- The code in views.py handles form input with POST or GET requests.  The user's input parameters by form POST are converted to a GET URL. The URL can be copied or downloaded to a file. With the URL, the plot(s) can be re-generated by GET method. 

* The code snippet is generated by calling a method in wormfunconn package. 

**Notes:**

- The form values are validated both client-side and server-side. The client side is through HTML form by defining Django form widgets. The server side validation uses Django's MaxValueValidator, MinValueValidator methods. The client side validation has a nice feature -- restricting the range and step of values by defining step, min, max for form field.

- Django form fields are mainly rendered manually in template for customizing display features. 

- For some form fields, the help text can be displayed by clicking a help icon to show collapsible card. Help text for each field are defined in forms.py.

- Found small bugs in wormfunconn package for generating code snippet. Should be an easy fixing.

- The introduction displayed on home page is the output of a method in wormfunconn.

fkang-pu commented:
Failed functional test due to scipy version. I used newest version (1.8.0) on my laptop, but may need to lower to 1.7.3. scipy is required by latest version of wormfunconn package. Will look into details and run possible tests before coming up with an action plan.

jcmatese reviewed (approved):
Looks ok to me, if it behaves the way you and the requestor wish.  Added a couple questions that came to mind.

lparsons commented on neuronsimulator/templates/home.html:
```diff
@@ -1,23 +1,251 @@
+
+            // show or hide form fields based on stim_type, use inital values for fields with display=none
+            if ( stim_type_value  === 'rectangular') {
+                document.getElementById("div_duration").style.display='block';
```
Consider using constants for the element ids as well, for the same reason as the `stim_type_value`. The values could be defined in python and inserted into the HTML (and/or JavaScript) templates, *e.g.*:

`forms.py` (Python): 
```python
DIV_DURATION="div_duration"
```
`home.html` (HTML Template): 
```html
<div id="{{ DIV_DURATION }}"
```
`home.html` (JavaScript): 
```javascript
document.getElementById("{{ DIV_DURATION }").style.display='block';
```

fkang-pu replied:
see my comments above.

lparsons reviewed (changes requested):
This seems to clean up the code a bit, at least to my eye. Nice work there. It also seems to address the issues we've discussed with Francesco.

I made a few comments and suggestions inline, mostly around using variables instead of repeating strings in the code. That would generally make the code more robust and easier to maintain.

The main issue I would like to resolve would `scipy`. I don't believe that the version of scipy used (`1.8.0`) is compatible with the version of Python running on `neuron-simulator`. I think when I asked Sven to update Python the last time (perhaps for TraceBase), he was able to do that very quickly and we're now at 3.8 or 3.9, so asking him to update might be the quickest fix. Ultimately, I'd like to update the python that GitHub actions is using to match the one we have on the server to try and catch this type of thing.

fkang-pu commented on neuronsimulator/forms.py:
```diff
@@ -1,110 +1,121 @@
+
+    STRAIN_CHOICES = [
+        ("wild type", "wild type"),
+        ("mutant1", "mutant1"),
```
Just a placeholder for now. Francesco wanted to have this option, but he hasn't added any code to handle different strains.

jcmatese replied:
OK, but might want to set a reminder on your "punch-list" because I imagine you will want that squared away before production/publication.

fkang-pu commented on neuronsimulator/tests.py:
```diff
@@ -33,31 +40,30 @@ def test_home_view(self):
+            "stim_type": "rectangular",
+            "stim_neu_id": "AVAL",
+            "resp_neu_ids": ["AVAL", "AVAR", "ASEL", "AWAL"],
             "nt": 1000,
```
The latest version of wormfunconn package can handle such case, output responses as well as a message "The activity of the stimulated neuron {stim_neu_id} is the activity set as stimulus".  So I modified my code in utils.py and forms.py accordingly.

fkang-pu commented on neuronsimulator/utils.py:
```diff
@@ -0,0 +1,320 @@
+            labels = []
+            msg = None
+
+        # if resp_neu_ids is None, the repsonses are for all neurons, need to filter based on top_n ranks
```
fixed now.

fkang-pu commented on requirements/common.txt:
```diff
@@ -1,5 +1,6 @@
 django-validators==1.0.1
-plotly==5.2.1
+plotly==5.6.0
+scipy==1.8.0
```
Updated Python to 3.9.

fkang-pu commented on requirements/dev.txt:
```diff
@@ -1,9 +1,9 @@
+black==21.12b0
+isort==5.10.1
+pylint==2.12.2
+mypy==0.931
```
Updated the versions as suggested. Thanks.

fkang-pu commented on .htmlhintrc:
```diff
@@ -13,7 +13,7 @@
   "alt-require": true,
   "doctype-html5": true,
-  "id-class-value": "dash",
+  "id-class-value": false,
```
The id for rendering Django form field is in "id_{field_name}" format. I think it's better to set "id-class-value" to false. It's also the setting used for TraceBase.

fkang-pu commented on neuronsimulator/forms.py:
```diff
@@ -1,110 +1,114 @@
+        ("delta", "delta"),
+        ("sinusoidal", "sinusoidal"),
+        ("realistic", "realistic"),
+    ]
```
I added a few methods to utils.py to get stim_types, and all optional form fields and attrs. Modified code in forms.py and home.html  to make code simpler and more robust.

fkang-pu commented on neuronsimulator/templates/base.html:
```diff
@@ -22,13 +33,28 @@
         </div>
 
-        <!-- Use Bootstrap JS -->
-        <script src="{% static 'bootstrap-5.0.1-dist/js/bootstrap.bundle.min.js' %}"></script>
```
Removed all unused css and JS scripts.

fkang-pu commented on neuronsimulator/templates/home.html:
```diff
@@ -1,23 +1,251 @@
+            let sel_option = document.getElementById("id_stim_type").selectedOptions;
+
+            // show or hide form fields based on stim_type, use inital values for fields with display=none
+            if ( stim_type_value  === 'rectangular') {
```
see my comments above.

fkang-pu commented:
I modified code in test.py due to changes made to utils.py and forms.py. After testing, I feel it's better to validate number fields in HTML form by setting min. and/or max. values. It seems unnecessary to use Django's validators (MinValueValidator and MaxValueValidator) for some fields, so I removed them.  Now I try to get form attrs. by calling FunctionalAtlas class methods if possible, and use them in Django form, template, and JavaScript.

lparsons commented on neuronsimulator/utils.py:
```diff
@@ -11,11 +11,78 @@
     capture error in each step and then log it to a dictionary (app_error_dict) in output
     """
 
+    @classmethod
+    def get_stim_type_list(cls):
+        """
+        get the list of stim_types defined by wormfunconn package
+        """
+        stim_type_list = ["rectangular", "delta", "sinusoidal", "realistic"]
+        return stim_type_list
```
NB. Nice idea to encapsulate these into some functions, it seems to clean up the templates, etc. quite a bit. Would it make sense to make these part of the form class? On first glance, it seems to me like that would keep the form info all together in that class.

fkang-pu replied:
I thought about it last Friday and decided to leave it in utils.py. In addition to forms.py,  this method is also called by "get_form_opt_field_dict" method in the same class in utils.py.

lparsons reviewed (approved):
I like the reduction in repetition in this update. I think it makes it a lot easier to update if needed going forward.
